tasty/src/dotty/tools/tasty/TastyHeaderUnpickler.scala

```diff
@@ -0,0 +1,136 @@
+package dotty.tools.tasty
+
+import java.util.UUID
+
+import TastyFormat.{MajorVersion, MinorVersion, ExperimentalVersion, header}
+
+/**
+ * The Tasty Header consists of four fields:
+ * - uuid
+ *   - contains a hash of the sections of the TASTy file
+ * - majorVersion
+ *   - matching the TASTy format version that last broke backwards compatibility
+ * - minorVersion
+ *   - matching the TASTy format version that last broke forward compatibility
+ * - experimentalVersion
+ *   - 0 for final compiler version
+ *   - positive for between minor versions and forward compatibility
+ *     is broken since the previous stable version.
+ * - toolingVersion
+ *   - arbitrary string representing the tooling that produced the TASTy
+ */
+sealed abstract case class TastyHeader(
+  uuid: UUID,
+  majorVersion: Int,
+  minorVersion: Int,
+  experimentalVersion: Int,
+  toolingVersion: String // String could lead to a lot of duplication, perhaps cache headers?
+)
+
+class TastyHeaderUnpickler(reader: TastyReader) {
+  import TastyHeaderUnpickler._
+  import reader._
+
+  def this(bytes: Array[Byte]) = this(new TastyReader(bytes))
+
+  /** reads and verifies the TASTy version, extracting the UUID */
+  def readHeader(): UUID =
+    readFullHeader().uuid
+
+  /** reads and verifies the TASTy version, extracting the whole header */
+  def readFullHeader(): TastyHeader = {
+
+    for (i <- 0 until header.length)
+      check(readByte() == header(i), "not a TASTy file")
+    val fileMajor = readNat()
+    if (fileMajor <= 27) { // old behavior before `tasty-core` 3.0.0-M4
+      val fileMinor = readNat()
+      val signature = signatureString(expectedMinor, fileMajor, fileMinor.toString)
+      throw new UnpickleException(signature + backIncompatAddendum + myAddendum)
+    }
+    else {
+      val fileMinor = readNat()
+      val fileExperimental = readNat()
+      val toolingVersion = {
+        val length = readNat()
+        val start = currentAddr
+        val end = start + length
+        goto(end)
+        new String(bytes, start.index, length) // NOTE: new string object here (should probably cache as it leaks)
+      }
+
+      val validVersion = ( // inlined from `dotty.tools.tasty.TastyVersionFormatTest.TastyVersion.<:<`
```

Comment:
Why does the production code contain code inlined from the test code?

Reply:
This one I guess comes under premature optimisation, but the one in the test code is the "canonical" version, but is less efficient due to constructing objects with fields. So I could put the logic in one shared parameterised function and then the one in the test code can delegate to that.